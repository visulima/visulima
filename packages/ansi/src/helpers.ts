// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
const isBrowser = globalThis?.window?.document !== undefined;

export const isTerminalApp = !isBrowser && process.env.TERM_PROGRAM === "Apple_Terminal";
export const isWindows = !isBrowser && process.platform === "win32";
