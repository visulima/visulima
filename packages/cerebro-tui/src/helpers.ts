import cliTruncate from "cli-truncate";
import stringWidth from "string-width";
import wordwrap from "wordwrap";

/**
 * Applies padding to the left or the right of the string by repeating
 * a given char.
 *
 * The method is not same as `padLeft` or `padRight` from JavaScript STD lib,
 * since it repeats a char regardless of the max width.
 */
const applyPadding = (value: string, options: { paddingChar: string; paddingLeft?: number; paddingRight?: number }): string => {
    if (options.paddingLeft) {
        // eslint-disable-next-line no-param-reassign
        value = `${options.paddingChar.repeat(options.paddingLeft)}${value}`;
    }

    if (options.paddingRight) {
        // eslint-disable-next-line no-param-reassign
        value = `${value}${options.paddingChar.repeat(options.paddingRight)}`;
    }

    return value;
};

/**
 * Justify the columns to have the same width by filling
 * the empty slots with a padding char.
 *
 * Optionally, the column can be aligned left or right.
 */
export const justify = (
    columns: string[],
    options: {
        align?: "left" | "right";
        maxWidth: number;
        paddingChar?: string;
    },
): string[] => {
    const normalizedOptions = {
        align: "left" as const,
        paddingChar: " ",
        ...options,
    };

    return columns.map((column) => {
        const columnWidth = stringWidth(column);

        /**
         * Column is already same or greater than the maxWidth
         */
        if (columnWidth >= normalizedOptions.maxWidth) {
            return column;
        }

        /**
         * Fill empty space on the right
         */
        if (normalizedOptions.align === "left") {
            return applyPadding(column, {
                paddingChar: normalizedOptions.paddingChar,
                paddingRight: normalizedOptions.maxWidth - columnWidth,
            });
        }

        /**
         * Fill empty space on the left
         */
        return applyPadding(column, {
            paddingChar: normalizedOptions.paddingChar,
            paddingLeft: normalizedOptions.maxWidth - columnWidth,
        });
    });
};

/**
 * Wrap the text under the starting and the ending column.
 * The first line will start at 1st column. However, from
 * the 2nd line onwards, the columns before the start
 * column are filled with white space.
 */
export const wrap = (
    columns: string[],
    options: {
        endColumn: number;
        startColumn: number;
        trimStart?: boolean;
    },
): string[] => {
    const wrapper = wordwrap(options.startColumn, options.endColumn);

    if (options.trimStart) {
        return columns.map((column) => wrapper(column).trimStart());
    }

    return columns.map((column) => wrapper(column));
};

/**
 * Truncates the text after a certain width.
 */
export const truncate = (
    columns: string[],
    options: {
        maxWidth: number;
        position?: "end" | "middle" | "start";
        truncationChar?: string;
    },
): string[] =>
    columns.map((column) =>
        cliTruncate(column, options.maxWidth, {
            position: options.position ?? "end",
            truncationCharacter: options.truncationChar ?? "â€¦",
        }),
    );
