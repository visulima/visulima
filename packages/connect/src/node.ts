import type { IncomingMessage, ServerResponse } from "node:http";
import type { AnyZodObject, ZodObject } from "zod";

import withZod from "./adapter/with-zod";
import type { Route } from "./router";
import { Router } from "./router";
import type {
    FindResult,
    FunctionLike,
    HandlerOptions,
    HttpMethod,
    Nextable,
    RouteMatch,
    RouteShortcutMethod,
    RoutesExtendedRequestHandler,
    ValueOrPromise,
} from "./types";

const onNoMatch = async (request: IncomingMessage, response: ServerResponse) => {
    response.statusCode = 404;
    response.end(request.method === "HEAD" ? undefined : `Route ${request.method} ${request.url} not found`);
};

const onError = async (error: unknown, _request: IncomingMessage, response: ServerResponse) => {
    response.statusCode = 500;

    console.error(error);

    response.end("Internal Server Error");
};

export const getPathname = (url: string): string => {
    const queryIndex = url.indexOf("?");

    return queryIndex === -1 ? url : url.slice(0, Math.max(0, queryIndex));
};

export type RequestHandler<Request extends IncomingMessage, Response extends ServerResponse> = (request: Request, response: Response) => ValueOrPromise<void>;

export class NodeRouter<
    Request extends IncomingMessage = IncomingMessage,
    Response extends ServerResponse = ServerResponse,
    Schema extends AnyZodObject = ZodObject<never>,
> {
    private readonly onError: (
        error: unknown,
        ...arguments_: Parameters<RoutesExtendedRequestHandler<Request, Response, Response, Route<Nextable<FunctionLike>>[]>>
    ) => ReturnType<RoutesExtendedRequestHandler<Request, Response, Response, Route<Nextable<FunctionLike>>[]>>;

    private readonly onNoMatch: RoutesExtendedRequestHandler<Request, Response, Response, Route<Nextable<FunctionLike>>[]>;

    private router = new Router<RequestHandler<Request, Response>>();

    public all: RouteShortcutMethod<this, Schema, RequestHandler<Request, Response>> = this.add.bind(this, "");

    public delete: RouteShortcutMethod<this, Schema, RequestHandler<Request, Response>> = this.add.bind(this, "DELETE");

    public get: RouteShortcutMethod<this, Schema, RequestHandler<Request, Response>> = this.add.bind(this, "GET");

    public head: RouteShortcutMethod<this, Schema, RequestHandler<Request, Response>> = this.add.bind(this, "HEAD");

    public patch: RouteShortcutMethod<this, Schema, RequestHandler<Request, Response>> = this.add.bind(this, "PATCH");

    public post: RouteShortcutMethod<this, Schema, RequestHandler<Request, Response>> = this.add.bind(this, "POST");

    public put: RouteShortcutMethod<this, Schema, RequestHandler<Request, Response>> = this.add.bind(this, "PUT");

    public constructor(options: HandlerOptions<RoutesExtendedRequestHandler<Request, Response, Response, Route<Nextable<FunctionLike>>[]>> = {}) {
        this.onNoMatch = options.onNoMatch ?? onNoMatch;
        this.onError = options.onError ?? onError;
    }

    private add(
        method: HttpMethod | "",
        routeOrFunction: Nextable<RequestHandler<Request, Response>> | RouteMatch,
        zodOrRouteOrFunction?: Nextable<RequestHandler<Request, Response>> | RouteMatch | Schema,
        ...fns: Nextable<RequestHandler<Request, Response>>[]
    ) {
        if (typeof routeOrFunction === "string" && typeof zodOrRouteOrFunction === "function") {
            // eslint-disable-next-line no-param-reassign
            fns = [zodOrRouteOrFunction];
        } else if (typeof zodOrRouteOrFunction === "object") {
            if (typeof routeOrFunction === "function") {
                // eslint-disable-next-line no-param-reassign
                fns = [withZod<Request, Response, Nextable<RequestHandler<Request, Response>>, Schema>(zodOrRouteOrFunction as Schema, routeOrFunction)];
            } else {
                // eslint-disable-next-line no-param-reassign
                fns = fns.map((function_) =>
                    withZod<Request, Response, Nextable<RequestHandler<Request, Response>>, Schema>(zodOrRouteOrFunction as Schema, function_),
                );
            }
        } else if (typeof zodOrRouteOrFunction === "function") {
            // eslint-disable-next-line no-param-reassign
            fns = [zodOrRouteOrFunction];
        }

        this.router.add(method, routeOrFunction, ...fns);

        return this;
    }

    // eslint-disable-next-line class-methods-use-this
    private prepareRequest(request: Request & { params?: Record<string, unknown> }, findResult: FindResult<RequestHandler<Request, Response>>) {
        request.params = {
            ...findResult.params,
            ...request.params, // original params will take precedence
        };
    }

    public clone(): NodeRouter<Request, Response, Schema> {
        const r = new NodeRouter<Request, Response, Schema>({ onError: this.onError, onNoMatch: this.onNoMatch });

        r.router = this.router.clone();

        return r;
    }

    public handler(): (request: Request, response: Response) => Promise<void> {
        const { routes } = this.router as Router<FunctionLike>;

        return async (request: Request, response: Response) => {
            // eslint-disable-next-line unicorn/no-array-callback-reference,unicorn/no-array-method-this-argument
            const result = this.router.find(request.method as HttpMethod, getPathname(request.url as string));

            this.prepareRequest(request, result);

            try {
                await (result.fns.length === 0 || result.middleOnly ? this.onNoMatch(request, response, routes) : Router.exec(result.fns, request, response));
            } catch (error) {
                await this.onError(error, request, response, routes);
            }
        };
    }

    public async run(request: Request, response: Response): Promise<unknown> {
        // eslint-disable-next-line unicorn/no-array-callback-reference,unicorn/no-array-method-this-argument
        const result = this.router.find(request.method as HttpMethod, getPathname(request.url as string));

        if (result.fns.length === 0) {
            return;
        }

        this.prepareRequest(request, result);

        // eslint-disable-next-line consistent-return
        return await Router.exec(result.fns, request, response);
    }

    public use(
        base: Nextable<RequestHandler<Request, Response>> | NodeRouter<Request, Response, Schema> | RouteMatch,
        ...fns: (Nextable<RequestHandler<Request, Response>> | NodeRouter<Request, Response, Schema>)[]
    ): this {
        if (typeof base === "function" || base instanceof NodeRouter) {
            fns.unshift(base);
            // eslint-disable-next-line no-param-reassign
            base = "/";
        }
        this.router.use(base, ...fns.map((function_) => (function_ instanceof NodeRouter ? function_.router : function_)));

        return this;
    }
}

export const createRouter = <
    Request extends IncomingMessage,
    Response extends ServerResponse,
    Schema extends AnyZodObject = ZodObject<{ body?: AnyZodObject; headers?: AnyZodObject; query?: AnyZodObject }>,
>(
    options: HandlerOptions<RoutesExtendedRequestHandler<Request, Response, Response, Route<Nextable<FunctionLike>>[]>> = {},
): NodeRouter<Request, Response, Schema> => new NodeRouter<Request, Response, Schema>(options);
