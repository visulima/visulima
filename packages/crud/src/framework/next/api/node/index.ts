import type { NextApiRequest, NextApiResponse } from "next";

import baseHandler from "../../../../base-crud-handler";
import type {
    Adapter, ExecuteHandler, HandlerOptions, ParsedQueryParameters,
} from "../../../../types";

async function handler<
    T,
    Q extends ParsedQueryParameters = any,
    R extends NextApiRequest = NextApiRequest,
    Response extends NextApiResponse = NextApiResponse,
    M extends string = string,
>(adapter: Adapter<T, Q>, options?: HandlerOptions<M, R, Response>): Promise<ExecuteHandler<R, Response>> {
    return baseHandler<R, Response, T, Q, M>(
        async (response, responseConfig) => {
            response.status(responseConfig.status).send(responseConfig.data);
        },
        async (response) => {
            response.end();
        },
        adapter,
        options,
    );
}

export default handler;
