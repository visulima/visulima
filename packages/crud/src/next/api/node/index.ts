import type { NextApiRequest, NextApiResponse } from "next";

import baseHandler from "../../../base-crud-handler";
import type { Adapter, ExecuteHandler, HandlerOptions, ParsedQueryParameters } from "../../../types.d";

const handler = async <
    T,
    Q extends ParsedQueryParameters = any,
    R extends NextApiRequest = NextApiRequest,
    Response extends NextApiResponse = NextApiResponse,
    M extends string = string,
>(
    adapter: Adapter<T, Q>,
    options?: HandlerOptions<M>,
): Promise<ExecuteHandler<R, Response>> =>
    await baseHandler<R, Response, T, Q, M>(
        async (response, responseConfig) => {
            response.status(responseConfig.status).send(responseConfig.data);
        },
        async (response) => {
            response.end();
        },
        adapter,
        options,
    );

export default handler;
