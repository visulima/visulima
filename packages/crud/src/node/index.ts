import type { IncomingMessage, ServerResponse } from "node:http";

import baseHandler from "../base-crud-handler";
import type {
    Adapter, ExecuteHandler, HandlerOptions, ParsedQueryParameters,
} from "../types.d";

async function handler<R extends IncomingMessage, Response extends ServerResponse, T, Q extends ParsedQueryParameters = any, M extends string = string>(
    adapter: Adapter<T, Q>,
    options?: HandlerOptions<M>,
): Promise<ExecuteHandler<R, Response>> {
    return baseHandler<R, Response, T, Q, M>(
        async (response, responseConfig) => {
            response.setHeader("content-type", "application/json; charset=utf-8");

            response.statusCode = responseConfig.status;
            response.end(JSON.stringify(responseConfig.data, null, 2));
        },
        async (response) => {
            (response as Response).end();
        },
        adapter,
        options,
    );
}

export default handler;
