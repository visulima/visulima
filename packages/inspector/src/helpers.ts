import type { Inspect, Options } from "./types";

const ansiColors = {
    // 21-29 are resets for 1-9
    black: ["30", "39"],
    blue: ["34", "39"],
    bold: ["1", "22"],
    brightblack: ["30;1", "39"],
    brightblue: ["34;1", "39"],
    brightcyan: ["36;1", "39"],
    brightgreen: ["32;1", "39"],
    // 10-20 are fonts
    brightmagenta: ["35;1", "39"],
    brightred: ["31;1", "39"],
    brightwhite: ["37;1", "39"],
    brightyellow: ["33;1", "39"],
    cyan: ["36", "39"],
    dim: ["2", "22"],
    green: ["32", "39"],
    grey: ["90", "39"],

    hidden: ["8", "28"],
    // 5 & 6 are blinking
    inverse: ["7", "27"],
    italic: ["3", "23"],
    magenta: ["35", "39"],
    red: ["31", "39"],
    strike: ["9", "29"],
    underline: ["4", "24"],
    white: ["37", "39"],

    yellow: ["33", "39"],
} as const;

const styles: Record<string, keyof typeof ansiColors> = {
    bigint: "yellow",
    boolean: "yellow",
    date: "magenta",
    null: "bold",
    number: "yellow",
    regexp: "red",
    special: "cyan",
    string: "green",
    symbol: "green",
    undefined: "grey",
} as const;

export const truncator = "â€¦";

type AnsiValue = (typeof ansiColors)[keyof typeof ansiColors];

function colorise<S extends string>(value: S, styleType: string): string {
    const color: AnsiValue | undefined = ansiColors[styles[styleType]] || ansiColors[styleType as keyof typeof ansiColors] || "";

    if (!color) {
        return String(value) as S;
    }

    return `\u001B[${color[0]}m${String(value) as S}\u001B[${color[1]}m`;
}

export function normaliseOptions(
    {
        breakLength = Number.POSITIVE_INFINITY,
        colors = false,
        customInspect = true,
        depth = 2,
        maxArrayLength = Number.POSITIVE_INFINITY,
        seen = [],
        showHidden = false,
        showProxy = false,
        stylize = String,

        truncate = Number.POSITIVE_INFINITY,
    }: Partial<Options> = {},
    inspect: Inspect,
): Options {
    const options = {
        breakLength: Number(breakLength),
        colors: Boolean(colors),
        customInspect: Boolean(customInspect),
        depth: Number(depth),
        inspect,
        maxArrayLength: Number(maxArrayLength),
        seen,
        showHidden: Boolean(showHidden),
        showProxy: Boolean(showProxy),
        stylize,
        truncate: Number(truncate),
    };
    if (options.colors) {
        options.stylize = colorise;
    }
    return options;
}

function isHighSurrogate(char: string): boolean {
    return char >= "\uD800" && char <= "\uDBFF";
}

export function truncate(string: number | string, length: number, tail: typeof truncator = truncator) {
    string = String(string);
    const tailLength = tail.length;
    const stringLength = string.length;
    if (tailLength > length && stringLength > tailLength) {
        return tail;
    }
    if (stringLength > length && stringLength > tailLength) {
        let end = length - tailLength;
        if (end > 0 && isHighSurrogate(string[end - 1])) {
            end -= 1;
        }
        return `${string.slice(0, end)}${tail}`;
    }
    return string;
}

export function inspectList(list: ArrayLike<unknown>, options: Options, inspectItem?: Inspect, separator = ", "): string {
    inspectItem = inspectItem || options.inspect;
    const size = list.length;
    if (size === 0) return "";
    const originalLength = options.truncate;
    let output = "";
    let peek = "";
    let truncated = "";
    for (let index = 0; index < size; index += 1) {
        const last = index + 1 === list.length;
        const secondToLast = index + 2 === list.length;
        truncated = `${truncator}(${list.length - index})`;
        const value = list[index];

        // If there is more than one remaining we need to account for a separator of `, `
        options.truncate = originalLength - output.length - (last ? 0 : separator.length);
        const string = peek || inspectItem(value, options) + (last ? "" : separator);
        const nextLength = output.length + string.length;
        const truncatedLength = nextLength + truncated.length;

        // If this is the last element, and adding it would
        // take us over length, but adding the truncator wouldn't - then break now
        if (last && nextLength > originalLength && output.length + truncated.length <= originalLength) {
            break;
        }

        // If this isn't the last or second to last element to scan,
        // but the string is already over length then break here
        if (!last && !secondToLast && truncatedLength > originalLength) {
            break;
        }

        // Peek at the next string to determine if we should
        // break early before adding this item to the output
        peek = last ? "" : inspectItem(list[index + 1], options) + (secondToLast ? "" : separator);

        // If we have one element left, but this element and
        // the next takes over length, the break early
        if (!last && secondToLast && truncatedLength > originalLength && nextLength + peek.length > originalLength) {
            break;
        }

        output += string;

        // If the next element takes us to length -
        // but there are more after that, then we should truncate now
        if (!last && !secondToLast && nextLength + peek.length >= originalLength) {
            truncated = `${truncator}(${list.length - index - 1})`;
            break;
        }

        truncated = "";
    }
    return `${output}${truncated}`;
}

function quoteComplexKey(key: string): string {
    if (/^[a-z_]\w*$/i.test(key)) {
        return key;
    }
    return JSON.stringify(key)
        .replaceAll("'", "\\'")
        .replaceAll('\\"', '"')
        .replaceAll(/(^"|"$)/g, "'");
}

export function inspectProperty([key, value]: [unknown, unknown], options: Options): string {
    options.truncate -= 2;
    if (typeof key === "string") {
        key = quoteComplexKey(key);
    } else if (typeof key !== "number") {
        key = `[${options.inspect(key, options)}]`;
    }
    options.truncate -= (key as string).length;
    value = options.inspect(value, options);
    return `${key}: ${value}`;
}
