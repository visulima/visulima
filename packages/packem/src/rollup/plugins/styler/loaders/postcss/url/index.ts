import path from "node:path";

import type { Declaration, PluginCreator } from "postcss";
import type { Node, ParsedValue } from "postcss-value-parser";
import valueParser from "postcss-value-parser";

import { isAbsolutePath,normalizePath } from "../../../utils/path";
import { mm } from "../../../utils/sourcemap";
import { dataURIRe,firstExtRe as firstExtensionRe } from "../common";
import generateName from "./generate";
import inlineFile from "./inline";
import type { UrlFile,UrlResolve } from "./resolve";
import resolveDefault from "./resolve";
import { isDeclWithUrl,walkUrls } from "./utils";

const name = "styles-url";
const placeholderHashDefault = "assets/[name]-[hash][extname]";
const placeholderNoHashDefault = "assets/[name][extname]";

/** URL handler options */
export interface UrlOptions {
    /**
     * Aliases for URL paths.
     * Overrides the global `alias` option.
     * - ex.: `{"foo":"bar"}`
     */
    alias?: Record<string, string>;
    /**
     * Directory path for outputted CSS assets,
     * which is not included into resulting URL
     * @default "."
     */
    assetDir?: string;
    /**
     * Enable/disable name generation with hash for outputted CSS assets
     * or provide your own placeholder with the following blocks:
     * - `[extname]`: The file extension of the asset including a leading dot, e.g. `.png`.
     * - `[ext]`: The file extension without a leading dot, e.g. `png`.
     * - `[hash(:<num>)]`: A hash based on the name and content of the asset (with optional length).
     * - `[name]`: The file name of the asset excluding any extension.
     *
     * Forward slashes / can be used to place files in sub-directories.
     * @default "assets/[name]-[hash][extname]" ("assets/[name][extname]" if false)
     */
    hash?: boolean | string;
    /**
     * Inline files instead of copying
     * @default true for `inject` mode, otherwise false
     */
    inline?: boolean;
    /**
     * Public Path for URLs in CSS files
     * @default "./"
     */
    publicPath?: string;
    /**
     * Provide custom resolver for URLs
     * in place of the default one
     */
    resolve?: UrlResolve;
}

const plugin: PluginCreator<UrlOptions> = (options = {}) => {
    const inline = options.inline ?? false;
    const publicPath = options.publicPath ?? "./";
    const assetDir = options.assetDir ?? ".";
    const resolve = options.resolve ?? resolveDefault;
    const alias = options.alias ?? {};
    const placeholder = options.hash ?? true ? (typeof options.hash === "string" ? options.hash : placeholderHashDefault) : placeholderNoHashDefault;

    return {

        async Once(css, { result: res }) {
            if (!css.source?.input.file) {
                return;
            }

            const { file } = css.source.input;
            const map = mm(css.source.input.map?.text).resolve(path.dirname(file)).toConsumer();

            const urlList: {
                basedirs: Set<string>;
                decl: Declaration;
                node: Node;
                parsed: ParsedValue;
                url: string;
            }[] = [];

            const imported = new Set(res.messages.filter((message) => message.type === "dependency").map((message) => message.file as string));

            css.walkDecls((decl) => {
                if (!isDeclWithUrl(decl)) {
                    return;
                }

                const parsed = valueParser(decl.value);
                walkUrls(parsed, (url, node) => {
                    // Resolve aliases
                    for (const [from, to] of Object.entries(alias)) {
                        if (url !== from && !url.startsWith(`${from}/`)) {
                            continue;
                        }

                        url = normalizePath(to) + url.slice(from.length);
                    }

                    // Empty URL
                    if (!node || url.length === 0) {
                        decl.warn(res, `Empty URL in \`${decl.toString()}\``);
                        return;
                    }

                    // Skip Data URI
                    if (dataURIRe.test(url)) {
                        return;
                    }

                    // Skip Web URLs
                    if (!isAbsolutePath(url)) {
                        try {
                            new URL(url);
                            return;
                        } catch {
                            // Is not a Web URL, continuing
                        }
                    }

                    const basedirs = new Set<string>();

                    // Use PostCSS imports
                    if (decl.source?.input.file && imported.has(decl.source.input.file)) {
                        basedirs.add(path.dirname(decl.source.input.file));
                    }

                    // Use SourceMap
                    if (decl.source?.start) {
                        const pos = decl.source.start;
                        const realPos = map?.originalPositionFor(pos);
                        const basedir = realPos?.source && path.dirname(realPos.source);

                        if (basedir) {
                            basedirs.add(path.normalize(basedir));
                        }
                    }

                    // Use current file
                    basedirs.add(path.dirname(file));

                    urlList.push({ basedirs, decl, node, parsed, url });
                });
            });

            const usedNames = new Map<string, string>();

            for await (const { basedirs, decl, node, parsed, url } of urlList) {
                let resolved: UrlFile | undefined;
                for await (const basedir of basedirs) {
                    try {
                        if (!resolved) {
                            resolved = await resolve(url, basedir);
                        }
                    } catch {
                        /* noop */
                    }
                }

                if (!resolved) {
                    decl.warn(res, `Unresolved URL \`${url}\` in \`${decl.toString()}\``);
                    continue;
                }

                const { from, source, urlQuery } = resolved;

                if (!(source instanceof Uint8Array) || typeof from !== "string") {
                    decl.warn(res, `Incorrectly resolved URL \`${url}\` in \`${decl.toString()}\``);
                    continue;
                }

                res.messages.push({ file: from, plugin: name, type: "dependency" });

                if (inline) {
                    node.type = "string";
                    node.value = inlineFile(from, source);
                } else {
                    const unsafeTo = normalizePath(generateName(placeholder, from, source));
                    let to = unsafeTo;

                    // Avoid file overrides
                    const hasExtension = firstExtensionRe.test(unsafeTo);
                    for (let index = 1; usedNames.has(to) && usedNames.get(to) !== from; index++) {
                        to = hasExtension ? unsafeTo.replace(firstExtensionRe, `${index}$1`) : `${unsafeTo}${index}`;
                    }

                    usedNames.set(to, from);

                    node.type = "string";
                    node.value = publicPath + (/[/\\]$/.test(publicPath) ? "" : "/") + path.basename(to);

                    if (urlQuery) {
                        node.value += urlQuery;
                    }

                    to = normalizePath(assetDir, to);
                    res.messages.push({ plugin: name, source, to, type: "asset" });
                }

                decl.value = parsed.toString();
            }
        },
        postcssPlugin: name,
    };
};

plugin.postcss = true;

export default plugin;
