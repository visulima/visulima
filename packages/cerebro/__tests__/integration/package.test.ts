import { fileURLToPath } from "node:url";

import { dirname, join } from "@visulima/path";
import stripAnsi from "strip-ansi";
import { describe, expect, it } from "vitest";

import { esc, execScriptSync } from "../helpers";

const strip = (string: string): string => esc(stripAnsi(string)).replaceAll("\r\n", "\n").trimEnd();

describe("usage `@visulima/cerebro` npm package", () => {
    it(`should work as CommonJS package`, () => {
        expect.assertions(1);

        const filename = join(dirname(fileURLToPath(import.meta.url)), "../..", "__fixtures__/package/cjs/test.cjs");

        const received = execScriptSync(filename);

        expect(strip(received)).toMatchSnapshot();
    });

    it(`should work as ESM package`, async () => {
        expect.assertions(1);

        const filename = join(dirname(fileURLToPath(import.meta.url)), "../..", "__fixtures__/package/mjs/test.mjs");

        const received = execScriptSync(filename);

        expect(strip(received)).toMatchSnapshot();
    });
});
