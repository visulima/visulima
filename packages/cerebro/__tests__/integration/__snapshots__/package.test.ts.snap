// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`usage \`@visulima/cerebro\` npm package > should work as CommonJS package 1`] = `
"cerebro
 Usage 

   cerebro <command> [positional arguments] [options]                            

 Available Commands 

   version  [v, V]                          Output the version number                                 
   help                                                                                               

 Command Options 

   --group string                           Display only the specified group      

 Global Options 

   --verbose boolean                        Turn on verbose output                
   --debug boolean                          Turn on debugging output              
   -h, --help boolean                       Print out helpful usage information   
   -q, --quiet boolean                      Silence output                        
   -V, --version boolean                    Print version info                    
   --no-color boolean                       Turn off colored output               
   --color boolean                          Force colored output                  

Run "cerebro help <command>" or "cerebro <command> --help" for help with a specific command."
`;

exports[`usage \`@visulima/cerebro\` npm package > should work as ESM package 1`] = `
"cerebro
 Usage 

   cerebro <command> [positional arguments] [options]                            

 Available Commands 

   version  [v, V]                          Output the version number                                 
   help                                                                                               

 Command Options 

   --group string                           Display only the specified group      

 Global Options 

   --verbose boolean                        Turn on verbose output                
   --debug boolean                          Turn on debugging output              
   -h, --help boolean                       Print out helpful usage information   
   -q, --quiet boolean                      Silence output                        
   -V, --version boolean                    Print version info                    
   --no-color boolean                       Turn off colored output               
   --color boolean                          Force colored output                  

Run "cerebro help <command>" or "cerebro <command> --help" for help with a specific command."
`;
